File: app/src/main/java/com/example/messenger/MainActivity.java
package com.example.messenger;

import androidx.appcompat.app.AppCompatActivity;
import androidx.lifecycle.ViewModelProvider;

import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.widget.Button;
import android.widget.EditText;
import android.widget.TextView;
import android.widget.Toast;

public class MainActivity extends AppCompatActivity {

    private static final String TAG = "MainActivity";

    private EditText editTextTextEmail;
    private EditText editTextPassword;
    private Button buttonLogin;
    private TextView textViewForgotPassword;
    private TextView textViewRegister;

    private MainViewModel viewModel;
    private final Context context = this;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        initViews();
        viewModel = new ViewModelProvider(this).get(MainViewModel.class);
        observeViewModel();
        setupClickListener();
    }

    private void observeViewModel() {
        viewModel.getError().observe(this, error -> {
                    if (error != null) {
                        Toast.makeText(this, error, Toast.LENGTH_SHORT).show();
                    }
                }
        );

        viewModel.getUser().observe(this, user -> {
            if (user != null) {
                Intent intent = UsersActivity.newIntent(context);
                startActivity(intent);
                finish();
            }
        });
    }

    private void setupClickListener() {
        buttonLogin.setOnClickListener(l -> {
            String email = getTrimmedValue(editTextTextEmail);
            String password = getTrimmedValue(editTextPassword);
            viewModel.login(email, password);
        });
        textViewRegister.setOnClickListener(l -> {
            Intent intent = RegistrationActivity.newIntent(context);
            startActivity(intent);
        });
        textViewForgotPassword.setOnClickListener(l -> {
            String email = getTrimmedValue(editTextTextEmail);
            Intent intent = ResetPasswordActivity.newIntent(context, email);
            startActivity(intent);
        });
    }

    private String getTrimmedValue(EditText editText) {
        return editText.getText().toString().trim();
    }

    public static Intent newIntent(Context context) {
        return new Intent(context, MainActivity.class);
    }

    private void initViews() {
        editTextTextEmail = findViewById(R.id.editTextEmail);
        editTextPassword = findViewById(R.id.editTextPassword);
        buttonLogin = findViewById(R.id.buttonLogin);
        textViewForgotPassword = findViewById(R.id.textViewForgotPassword);
        textViewRegister = findViewById(R.id.textViewRegister);
    }
}

File: app/src/main/java/com/example/messenger/MainActivity.kt
package com.example.messenger

import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.widget.EditText
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.lifecycle.ViewModelProvider
import com.example.messenger.databinding.ActivityMainBinding

class MainActivity : AppCompatActivity() {

    private lateinit var binding: ActivityMainBinding
    private lateinit var viewModel: MainViewModel
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityMainBinding.inflate(layoutInflater)
        setContentView(binding.root)
        viewModel = ViewModelProvider(this)[MainViewModel::class.java]
        observeViewModel()
        setupClickListener()
    }

    private fun setupClickListener() {
        with(binding) {
            buttonLogin.setOnClickListener {
                val email = editTextEmail.getTrimmedValue()
                val password = editTextPassword.getTrimmedValue()
                viewModel.login(email, password)
            }
            textViewRegister.setOnClickListener {

            }
            textViewForgotPassword.setOnClickListener {
                val email = editTextEmail.getTrimmedValue()
                val intent = ResetPasswordActivity.newIntent(this@MainActivity, email)
                startActivity(intent)
            }
        }
    }

    private fun observeViewModel() {
        with(viewModel) {
            getError.observe(this@MainActivity) {
                it?.let { Toast.makeText(this@MainActivity, it, Toast.LENGTH_SHORT).show() }
            }
            getUser.observe(this@MainActivity) {
                it?.let {

                }
            }
        }
    }

    companion object {
        fun newIntent(context: Context) = Intent(context, MainActivity::class.java)
    }

    private fun EditText.getTrimmedValue() = this.text.toString().trim()
}